TPP.Persistence.MongoDB/Repos/BadgeMarketRepo.cs

```diff
@@ -0,0 +1,249 @@
+            if (species != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Species, species);
+            if (form != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Form, form);
+            if (source != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Source, source);
+            if (shiny == true)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Shiny, true);
+            else
+                filter &= Builders<BadgeBuyOffer>.Filter.Ne(b => b.Shiny, true);
+
+            return await BuyOfferCollection.Find(filter).ToListAsync();
+        }
+
+        public async Task<List<Badge>> FindAllBadgesForSale(string? userId, PkmnSpecies species, string? form, Badge.BadgeSource? source, bool? shiny)
+        {
+            return await _badgeRepo.FindAllForSaleByCustom(userId, species, form, source, shiny);
+        }
+
+        public async Task<BadgeBuyOffer> CreateBuyOffer(string userId, PkmnSpecies species, string? form, Badge.BadgeSource? source, bool? shiny, int price, int amount, Instant? createdAt = null)
+        {
+            BadgeBuyOffer buyOffer = new BadgeBuyOffer(
+                id: string.Empty,
+                userId: userId,
+                species: species,
+                form: form,
+                source: source,
+                shiny: shiny,
+                price: price,
+                amount: amount,
+                createdAt: createdAt ?? _clock.GetCurrentInstant()
+                );
+
+            await BuyOfferCollection.InsertOneAsync(buyOffer);
+            Debug.Assert(buyOffer.Id.Length > 0, "The MongoDB driver injected a generated ID");
+            return buyOffer;
+        }
+
+        public async Task<Badge> CreateSellOffer(string userId, PkmnSpecies species, string? form, Badge.BadgeSource? source, bool? shiny, int price)
+        {
+            List<Badge> notSellingOwnedByUser = await _badgeRepo.FindAllNotForSaleByCustom(userId, species, form, source, shiny);
+            Badge toSell = SortBySpecialness(notSellingOwnedByUser).First();
+            Badge selling = await _badgeRepo.SetBadgeSellPrice(toSell, price);
+            return selling;
+        }
+
+        public async Task DeleteBuyOffer(string userId, PkmnSpecies species, string? form, Badge.BadgeSource? source, bool? shiny, int amount)
+        {
+            List<BadgeBuyOffer> offers = await FindAllBuyOffers(userId, species, form, source, shiny);
+            if(offers.Count() < amount)
+                throw new ArgumentException(string.Format("Tried to cancel {0} offers but only {1} were found", amount, offers.Count));
+            
+            offers = offers.OrderByDescending(o => o.CreatedAt).ToList();
+            for(int i=0; i < amount; i++)
+            {
+                await BuyOfferCollection.FindOneAndDeleteAsync(o => o.Id == offers[i].Id);
+            }
```

Comment:
Could this possibly be batched? Feels like this should be 1 transaction. Same for DeleteSellOffer and possibly others